src/index.ts

```diff
@@ -329,105 +381,162 @@ function logout() {
-    }];
-    const applicationCreateParameters = {
-        availableToOtherTenants: false,
-        displayName: azureWebsiteName,
-        homepage,
-        identifierUris,
-        oauth2AllowImplicitFlow: true,
-        passwordCredentials: [{
-                endDate,
-                keyId: uuid.v1(),
-                startDate,
-                value: newServicePrincipalSecret
-            }
-        ],
-        replyUrls,
-        requiredResourceAccess
-    };
+    const newServicePrincipalSecret: string = uuid.v4();
+    let newServicePrincipal: ServicePrincipal;
     let objectId: string = '';
-    return graphClient.applications.create(applicationCreateParameters)
+
+    let servicePrincipalCreateParameters;
+    let graphClient: GraphRbacManagementClient;
+
+    options.tokenAudience = 'graph';
+    let graphClientPromise: Promise<GraphRbacManagementClient>;
+    if (usingServicePrincipal) {
+        graphClientPromise = loginWithServicePrincipalSecretWithAuthResponse(
+            program.servicePrincipalId,
+            program.servicePrincipalSecret,
+            program.domainId,
+            options)
+            .then((response: AuthResponse) => {
+                return new GraphRbacManagementClient(response.credentials, options.domain || '', baseUri);
+            });
+    } else {
+        graphClient = new GraphRbacManagementClient(new DeviceTokenCredentials(options), options.domain || '', baseUri);
+        graphClientPromise = Promise.resolve<GraphRbacManagementClient>(graphClient);
+    }
+
+    return graphClientPromise.then((client: GraphRbacManagementClient) => {
+        graphClient = client;
+        const startDate = new Date(Date.now());
+        let endDate = new Date(startDate.toISOString());
+        const m = momemt(endDate);
+        m.add(1, 'years');
+        endDate = new Date(m.toISOString());
+        const identifierUris = [ homepage ];
+        const replyUrls = [ homepage ];
+        // Allowing Graph API to sign in and read user profile for newly created application
+        const requiredResourceAccess = [{
+            resourceAccess: [
+                {
+                    // This guid represents Sign in and read user profile
+                    // http://www.cloudidentity.com/blog/2015/09/01/azure-ad-permissions-summary-table/
+                    id: '311a71cc-e848-46a1-bdf8-97ff7156d8e6',
+                    type: 'Scope'
+                }
+            ],
+            // This guid represents Directory Graph API ID
+            resourceAppId: '00000002-0000-0000-c000-000000000000'
+        }];
+        const applicationCreateParameters: ApplicationCreateParameters = {
+            availableToOtherTenants: false,
+            displayName: azureWebsiteName,
+            homepage,
+            identifierUris,
+            oauth2AllowImplicitFlow: true,
+            replyUrls,
+            requiredResourceAccess
+        };
+
+        applicationCreateParameters.passwordCredentials = [{
```

Comment:
if we are creating passwordCredentials everytime then we can merge it with applicationCreateParameters initialization